mmhuman3d/utils/geometry.py

```diff
@@ -391,3 +415,157 @@ def weak_perspective_projection(points, scale, translation):
+            'bbox_center']
+        img_w, img_h, _ = pred_camera['img_w'], pred_camera[
+            'img_h'], pred_camera['crop_res']
+        cam_sxy, cam_rotmat, cam_intrinsics = pred_camera[
+            'cam_sxy'], pred_camera['cam_rotmat'], pred_camera[
+                'cam_intrinsics']
+        if 'cam_t' in pred_camera:
+            cam_t = pred_camera['cam_t']
+        else:
+            cam_t = convert_to_full_img_cam(
+                pare_cam=cam_sxy,
+                bbox_height=bbox_scale * 200.,
+                bbox_center=bbox_center,
+                img_w=img_w,
+                img_h=img_h,
+                focal_length=cam_intrinsics[:, 0, 0],
+            )
+
+        pred_keypoints_2d = perspective_projection(
+            pred_joints, rotation=cam_rotmat, translation=cam_t)
+    return pred_keypoints_2d
+
+
+def convert_to_full_img_cam(pare_cam, bbox_height, bbox_center, img_w, img_h,
+                            focal_length):
+    # Converts weak perspective camera estimated by PARE in
+    # bbox coords to perspective camera in full image coordinates
+    # from https://arxiv.org/pdf/2009.06549.pdf
+    s, tx, ty = pare_cam[:, 0], pare_cam[:, 1], pare_cam[:, 2]
+    res = 224
+    r = bbox_height / res
+    tz = 2 * focal_length / (r * res * s)
+
+    cx = 2 * (bbox_center[:, 0] - (img_w / 2.)) / (s * bbox_height)
+    cy = 2 * (bbox_center[:, 1] - (img_h / 2.)) / (s * bbox_height)
+
+    if torch.is_tensor(pare_cam):
+        cam_t = torch.stack([tx + cx, ty + cy, tz], dim=-1)
+    else:
+        cam_t = np.stack([tx + cx, ty + cy, tz], axis=-1)
+
+    return cam_t
+
+
+def compute_twist_rotation(rotation_matrix, twist_axis):
+    '''
+    Compute the twist component of given rotation and twist axis
+    https://stackoverflow.com/questions/3684269/component-of-a-quaternion-rotation-around-an-axis
+    Parameters
+    ----------
+    rotation_matrix : Tensor (B, 3, 3,)
+        The rotation to convert
+    twist_axis : Tensor (B, 3,)
+        The twist axis
+    Returns
+    -------
+    Tensor (B, 3, 3)
+        The twist rotation
+    '''
+    quaternion = rotation_matrix_to_quaternion(rotation_matrix)
+
+    twist_axis = twist_axis / (
+        torch.norm(twist_axis, dim=1, keepdim=True) + 1e-9)
+
+    projection = torch.einsum('bi,bi->b', twist_axis,
+                              quaternion[:, 1:]).unsqueeze(-1) * twist_axis
+
+    twist_quaternion = torch.cat([quaternion[:, 0:1], projection], dim=1)
+    twist_quaternion = twist_quaternion / (
+        torch.norm(twist_quaternion, dim=1, keepdim=True) + 1e-9)
+
+    twist_rotation = quaternion_to_rotation_matrix(twist_quaternion)
+    twist_aa = quaternion_to_angle_axis(twist_quaternion)
+
+    twist_angle = torch.sum(
+        twist_aa, dim=1, keepdim=True) / torch.sum(
+            twist_axis, dim=1, keepdim=True)
+
+    return twist_rotation, twist_angle
+
+
+def quaternion_to_rotation_matrix(quat):
```

Comment:
We already have `quat_to_rotmat` and `rotmat_to_rot6d` in [transforms](https://github.com/open-mmlab/mmhuman3d/blob/main/mmhuman3d/utils/transforms.py).